homeassistant/util/hass_dict.py

```diff
@@ -0,0 +1,231 @@
+
+    def __getitem__(self, key: _Key[Any] | str, /) -> Any:
+        """Forward to super implementation."""
+        return super().__getitem__(key)
+
+    # ------
+    @overload  # type: ignore[override]
+    def __setitem__(self, key: HassEntryKey[_T], value: dict[str, _T], /) -> None: ...
+
+    @overload
+    def __setitem__(self, key: HassKey[_T], value: _T, /) -> None: ...
+
+    @overload
+    def __setitem__(self, key: str, value: Any, /) -> None: ...
+
+    def __setitem__(self, key: _Key[Any] | str, value: Any, /) -> None:
+        """Forward to super implementation."""
+        return super().__setitem__(key, value)
+
+    # ------
+    @overload  # type: ignore[override]
+    def setdefault(
+        self, key: HassEntryKey[_T], default: dict[str, _T], /
+    ) -> dict[str, _T]: ...
+
+    @overload
+    def setdefault(self, key: HassKey[_T], default: _T, /) -> _T: ...
+
+    @overload
+    def setdefault(self, key: str, default: None = None, /) -> Any | None: ...
+
+    @overload
+    def setdefault(self, key: str, default: Any, /) -> Any: ...
+
+    def setdefault(self, key: _Key[Any] | str, default: Any = None, /) -> Any:
+        """Forward to super implementation."""
+        return super().setdefault(key, default)
+
+    # ------
+    @overload  # type: ignore[override]
+    def get(self, key: HassEntryKey[_T], /) -> dict[str, _T] | None: ...
+
+    @overload
+    def get(self, key: HassEntryKey[_T], default: _U, /) -> dict[str, _T] | _U: ...
+
+    @overload
+    def get(self, key: HassKey[_T], /) -> _T | None: ...
+
+    @overload
+    def get(self, key: HassKey[_T], default: _U, /) -> _T | _U: ...
+
+    @overload
+    def get(self, key: str, /) -> Any | None: ...
+
+    @overload
+    def get(self, key: str, default: Any, /) -> Any: ...
+
+    def get(self, key: _Key[Any] | str, default: Any = None, /) -> Any:
+        """Forward to super implementation."""
+        return super().get(key, default)
+
+    # ------
+    @overload  # type: ignore[override]
+    def pop(self, key: HassEntryKey[_T], /) -> dict[str, _T]: ...
+
+    @overload
+    def pop(
+        self, key: HassEntryKey[_T], default: dict[str, _T], /
+    ) -> dict[str, _T]: ...
+
+    @overload
+    def pop(self, key: HassEntryKey[_T], default: _U, /) -> dict[str, _T] | _U: ...
+
+    @overload
+    def pop(self, key: HassKey[_T], /) -> _T: ...
+
+    @overload
+    def pop(self, key: HassKey[_T], default: _T, /) -> _T: ...
+
+    @overload
+    def pop(self, key: HassKey[_T], default: _U, /) -> _T | _U: ...
+
+    @overload
+    def pop(self, key: str, /) -> Any: ...
+
+    @overload
+    def pop(self, key: str, default: _U, /) -> Any | _U: ...
+
+    def pop(self, key: _Key[Any] | str, default: Any = None, /) -> Any:
+        """Forward to super implementation."""
+        return super().pop(key, default)
+
+
+if TYPE_CHECKING:
+
+    def test_hass_dict_typing() -> None:
+        """Test HassDict overloads work as intended.
```

Comment:
Is there another way this can be tested without having to put it here?

Reply:
The tests folder is excluded from mypy. We could check a subfolder, e.g. `tests/typing` or create a new toplevel one like `typing_tests`. However, personally I feel like this belongs more closely together which is why I put in the same file as the implementation.

It was a real help to validate all overloads are working correctly and everything is as it should be.

--
_I would also like to differentiate here that this is a `util` file not `core.py` or `helpers`._

Reply:
I think `tests/typing` or `typing_tests` would be a better especially if we want to have more of these since we still have to pay the cost of having them loaded in memory even if we aren't running the code at runtime.

Reply:
Fair point. Will move them.

Reply:
Moved it together with the `if TYPE_CHECKING` implementation to a stub file (`hass_dict.pyi`). That way it should no longer be loaded into memory while still being used by mypy / pyright.

Would it be ok to keep them in the stub file? `tests/typing` has the disadvantage that it might be picked up by pytest during test discovery. Additionally, most of our strict mypy config is ignored for tests, so that would need to be adjusted. Not sure I would want to create a new top level folder just for it either though.

Reply:
I think thats ok since it never going to be loaded at runtime